src/core/mesh/qgsmeshlayerrenderer.cpp

```diff
@@ -345,15 +413,223 @@ void QgsMeshLayerRenderer::renderScalarDataset()
+              }
+            }
+          }
+        }
+        else if ( classificationType == QgsColorRampShader::Type::Discrete )
+        {
+          QgsPointXY startPoint = lineStart;
+          QColor color = colorAt( shader.get(), valVertexStart );
+          Q_ASSERT( ! qgsDoubleNear( valDiff, 0.0 ) );
+          for ( int i = 0; i < colorRampItemList.size() - 1; ++i )
+          {
+            const QgsColorRampShader::ColorRampItem &item = colorRampItemList.at( i );
+            if ( !std::isnan( item.value ) )
+            {
+              double fraction = ( item.value - valVertexStart ) / valDiff;
+              if ( ( fraction > 0.0 ) && ( fraction < 1.0 ) )
+              {
+                QgsPointXY endPoint = fractionPoint( lineStart, lineEnd, fraction, 0.0 );
+                pen.setColor( color );
+                painter->setPen( pen );
+                painter->drawLine( startPoint.toQPointF(), endPoint.toQPointF() );
+                color = item.color;
+                startPoint = endPoint;
+              }
+            }
+          }
+          pen.setColor( colorAt( shader.get(), valVertexEnd ) );
+          painter->setPen( pen );
+          painter->drawLine( startPoint.toQPointF(), lineEnd.toQPointF() );
+        }
+        else if ( classificationType == QgsColorRampShader::Type::Interpolated )
+        {
+          QLinearGradient gradient( lineStart.toQPointF(), lineEnd.toQPointF() );
+          gradient.setColorAt( 0.0, colorAt( shader.get(), valVertexStart ) );
+          gradient.setColorAt( 1.0, colorAt( shader.get(), valVertexEnd ) );
+          gradient.setSpread( gradient.ReflectSpread );
+
+          Q_ASSERT( ! qgsDoubleNear( valDiff, 0.0 ) );
+
+          for ( int i = 0; i < colorRampItemList.size(); ++i )
+          {
+            const QgsColorRampShader::ColorRampItem &item = colorRampItemList.at( i );
+            if ( !std::isnan( item.value ) )
+            {
+              double fraction = ( item.value - valVertexStart ) / valDiff;
+              if ( ( fraction > 0.0 ) && ( fraction < 1.0 ) )
+              {
+                gradient.setColorAt( fraction, colorAt( shader.get(), item.value ) );
+              }
+            }
+            QBrush brush( gradient );
+            pen.setBrush( brush );
+            painter->setPen( pen );
+            painter->drawLine( lineStart.toQPointF(), lineEnd.toQPointF() );
+          }
+        }
+        else
+        {
+          // no other option possible
+          Q_ASSERT( false );
+        }
+      }
+    }
+  }
+  painter->restore();
+}
+
+QColor QgsMeshLayerRenderer::colorAt( QgsColorRampShader *shader, double val ) const
+{
+  int r, g, b, a;
+  if ( shader->shade( val, &r, &g, &b, &a ) )
+  {
+    return QColor( r, g, b, a );
+  }
+  return QColor();
+}
+
+QgsPointXY QgsMeshLayerRenderer::fractionPoint( const QgsPointXY &p1, const QgsPointXY &p2, double fraction, double overlayMagniture ) const
```

Comment:
Can we use QgsGeometryUtils::interpolatePointOnLine() instead?

the `overlayMagniture` argument seems to be always called with 0, do we need it?